### PR TITLE
🚀 Parsely: Eliminate unnecessary http requests for incremental engaged time tracking

### LINKED ISSUE #13354: Customizable conditional triggers
amp-analytics provides for a specific set of events that can cause analytics requests to be made. While this covers a lot of popular use cases, there are at least two imaginable ways it can not support what one is looking to do:

1. You might want a completely new kind of trigger event. For example, some amp-bind state value being reached.

2. You want to generally trigger and event, but hold it back in case some condition isn't met.

This feature request covers case #2. For example, say you have a `timer` trigger going off every 5 seconds. As part of the request on that `timer` trigger you send the value of the viewport using INCREMENTAL_ENGAGED_TIME. However, you do NOT want to capture this event if the value of INCREMENTAL_ENGAGED_TIME is zero. There should be a way to define the trigger and also define a boolean expression for whether or not it gets blocked.

### PR DESCRIPTION
**Pull request by cwisecarver**
This PR uses the 'enabled' feature for triggers to eliminate incremental engaged time heartbeats where the incremental engaged time is 0. 

- Adds a second parameter to getIncrementalEngagedTime, `reset`, which defaults to true. This parameter allows calculating incremental engaged time without resetting the counter.
- Adds an enabled check to the heartbeat trigger of the Parsely test page 
    - Do not reset the counter in the enabled check
    - Do reset the counter in the actual heartbeat URL.

Resolves #13354

### CODE REVIEW

**Comment by cwisecarver** [comment]
@rudygalfi This eliminates the heartbeats with a value of zero.

**Comment by cwisecarver** [comment]
Tests fail locally but they're not related to my changes. I imagine they're failing in master as well.

File: extensions/amp-analytics/0.1/activity-impl.js [hunk 1]
```diff
@@ -318,18 +318,22 @@ export class Activity {
     return this.activityHistory_.getTotalEngagedTime(secondsSinceStart);
   }
   /**
-   * Get the incremental engaged time since the last push and reset it.
+   * Get the incremental engaged time since the last push and reset it if asked.
    * @param {string} name
+   * @param {boolean} reset
```
Comment (elgreg) [inline extensions/amp-analytics/0.1/activity-impl.js]: Should this be a `{string}` since is passed in as 'true'? Alternatively pass in as `true`?

Comment (cwisecarver) [reply extensions/amp-analytics/0.1/activity-impl.js]: I changed it to string. IIRC the url replacement code doesn't like types that aren't strings.

Comment (cwisecarver) [reply extensions/amp-analytics/0.1/activity-impl.js]: Thanks for catching that @elgreg. Hopefully we can get this reviewed and merged soon.

**Comment by cwisecarver** [comment]
Tests passing now after a rebase on master @rudygalfi.

File: extensions/amp-analytics/0.1/activity-impl.js (hunk 1 above)
Comment (elgreg) [reply extensions/amp-analytics/0.1/activity-impl.js]: My pleasure! We're really looking forward to using this in our implementation at Slate!

File: extensions/amp-analytics/0.1/activity-impl.js [hunk 2]
```diff
@@ -318,18 +318,22 @@ export class Activity {
...
   }
   /**
-   * Get the incremental engaged time since the last push and reset it.
+   * Get the incremental engaged time since the last push and reset it if asked.
    * @param {string} name
+   * @param {string} reset
    * @return {number}
    */
-  getIncrementalEngagedTime(name = '') {
+  getIncrementalEngagedTime(name = '', reset = 'true') {
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/activity-impl.js]: I wonder if it is easier to just create another increamentalEnagedTime value with a different name and not adding the reset value. 
But I am good with adding `reset`. Please add test and documentation.

File: extensions/amp-analytics/0.1/vendors.js [hunk 3]
```diff
@@ -1422,13 +1422,22 @@ export const ANALYTICS_CONFIG = /** @type {!JsonObject} */ ({
         'ampid=${clientId(_parsely_visitor)}',
       'pageview': '${basePrefix}&action=pageview',
       'heartbeat': '${basePrefix}&action=heartbeat' +
-      '&tt=${totalEngagedTime}&inc=${incrementalEngagedTime(parsely-js)}',
+      '&tt=${totalEngagedTime}&inc=${incrementalEngagedTime(parsely-js,true)}',
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/vendors.js]: nit: Please make this change in separate PR, thanks.

Comment (cwisecarver) [reply extensions/amp-analytics/0.1/vendors.js]: Good call 👍

**Comment by cwisecarver** [comment]
I've replied to the review @zhouyx. The tests that are failing are unrelated to my changes. I've just rebased on master again, just now, and they are still failing locally.

File: extensions/amp-analytics/0.1/activity-impl.js (hunk 2 above)
Comment (cwisecarver) [reply extensions/amp-analytics/0.1/activity-impl.js]: I started off creating a separate function but I ended up doing this instead. My thinking was that since I would have to expose it in so many places adding the flag as a function argument was a cleaner implementation.

**Comment by cwisecarver** [comment]
I responded @zhouyx I think I got everything. Thanks for the review. There are 3 tests failing locally unrelated to my changes. I also rebased on master right before pushing.

File: extensions/amp-analytics/0.1/activity-impl.js [hunk 4]
```diff
@@ -318,18 +318,22 @@ export class Activity {
     return this.activityHistory_.getTotalEngagedTime(secondsSinceStart);
   }
   /**
-   * Get the incremental engaged time since the last push and reset it.
+   * Get the incremental engaged time since the last push and reset it if asked.
    * @param {string} name
+   * @param {string=} reset
```
Comment (lannka) [inline extensions/amp-analytics/0.1/activity-impl.js]: let's have it as a boolean, and do the string to boolean conversion outside.

File: spec/amp-var-substitutions.md [hunk 5]
```diff
@@ -1085,6 +1086,14 @@ Provides the total time (in seconds) the user has been engaged with the page sin
 * **amp-analytics variable**: `${totalEngagedTime}`
   * Example value: `36`
 
+#### Incremental Engaged Time
+
+Provides the time (in seconds) the user has been engaged with the page since the last time it was reset. It takes two arguments. The first is the name of the timer, the second is whether or not to reset it (defaults to true). Incremental engaged time will be 0 until the page first becomes visible. This variable requires the [amp-analytics](../extensions/amp-analytics/amp-analytics.md) extension to be present on the page.
```
Comment (zhouyx) [inline spec/amp-var-substitutions.md]: nit: It will be great to state that the second argument is optional.

**Comment by cwisecarver** [comment]
@lannka @zhouyx I think I've addressed all the issues.

File: src/service/url-replacements-impl.js [hunk 6]
```diff
@@ -510,9 +510,14 @@ export class GlobalVariableSource extends VariableSource {
...
-    this.setAsync('INCREMENTAL_ENGAGED_TIME', name => {
+    this.setAsync('INCREMENTAL_ENGAGED_TIME', (name, reset) => {
       return Services.activityForDoc(this.ampdoc).then(activity => {
-        return activity.getIncrementalEngagedTime(name);
+        if (reset === 'false') {
+          reset = false;
+        } else {
+          reset = true;
+        }
+        return activity.getIncrementalEngagedTime(name, reset);
```
Comment (lannka) [inline src/service/url-replacements-impl.js]: you can simplify this by:

```
return activity.getIncrementalEngagedTime(name, reset !== 'false');
```

**Comment by cwisecarver** [comment]
Should be fixed after the next build @zhouyx @lannka

**Comment by cwisecarver** [comment]
Rebased again. I'll check in in a few hours.

**Comment by cwisecarver** [comment]
@zhouyx @lannka Tests are finally passing.

**Comment by cschieffer** [comment]
Thank you to the Parsely team for prioritizing this work and for all the other folks who contributed. Having this is huge!

**Comment by cwisecarver** [comment]
@lannka Could you give me a timeline for testing and release? Some of our customers are very eager to know when they'll be able to take advantage of this. Thanks!!

**Comment by cschieffer** [comment]
Looks like this was deployed?

**Comment by cwisecarver** [comment]
@lannka @zhouyx Any idea when this will be deployed? I also noticed that the trigger for the heartbeat some how made it's way back into vendors.js. I must have rebased off the wrong branch. Here's a PR to remove it. It would be great to get that merged before this is released. https://github.com/ampproject/amphtml/pull/15131
